examples/training/textual_inversion.py

```diff
@@ -0,0 +1,787 @@
+        model, optimizer, train_dataloader, lr_scheduler
+    )
+
+    # We need to recalculate our total training steps as the size of the training dataloader may have changed.
+    num_update_steps_per_epoch = math.ceil(len(train_dataloader) / args.gradient_accumulation_steps)
+    if overrode_max_train_steps:
+        args.max_train_steps = args.num_epochs * num_update_steps_per_epoch
+    # Afterwards we recalculate our number of training epochs
+    args.num_epochs = math.ceil(args.max_train_steps / num_update_steps_per_epoch)
+
+    if args.push_to_hub:
+        repo = init_git_repo(args, at_init=True)
+
+    if accelerator.is_main_process:
+        run = os.path.split(__file__)[-1].split(".")[0]
+        accelerator.init_trackers(run)
+
+    global_step = 0
+    for epoch in range(args.num_epochs):
+        model.train()
+        progress_bar = tqdm(total=num_update_steps_per_epoch, disable=not accelerator.is_local_main_process)
+        progress_bar.set_description(f"Epoch {epoch}")
+        for step, batch in enumerate(train_dataloader):
+            with accelerator.accumulate(model):
+                # Convert images to latent space
+                latents = model.vae.encode(batch["pixel_values"]).sample().detach()
+                latents = latents * 0.18215
+
+                # Sample noise that we'll add to the latents
+                noise = torch.randn(latents.shape).to(latents.device)
+                bsz = latents.shape[0]
+                # Sample a random timestep for each image
+                timesteps = torch.randint(0, noise_scheduler.num_train_timesteps, (bsz,), device=latents.device).long()
+
+                # Add noise to t1`he latents according to the noise magnitude at each timestep
+                # (this is the forward diffusion process)
+                noisy_latents = noise_scheduler.add_noise(latents, noise, timesteps)
+
+                # get the text embedding for conditioning
+                input_ids = batch["input_ids"].reshape(bsz, -1)
+                encoder_hidden_states = model.text_encoder(input_ids)[0]
+
+                # Predict the noise residual
+                noise_pred = model.unet(noisy_latents, timesteps, encoder_hidden_states)["sample"]
+                loss = F.mse_loss(noise_pred, noise, reduction="none").mean([1, 2, 3]).mean()
+
+                # Compute coarse loss
+                # TODO: Choose better name for embedding_reg_weight
+                if args.embedding_reg_weight > 0:
+                    num_embeddings = 1  # TODO: generalize to multiple embeddings
+                    optimized = model.text_encoder.get_concept_embeddings()
+                    coarse = model.text_encoder.get_initializer_embeddings().clone().to(optimized.device)
+                    coarse_loss = (optimized - coarse) @ (optimized - coarse).T / num_embeddings
+
+                    loss += args.embedding_reg_weight * coarse_loss
```

Comment:
Need to discuss with authors if this is needed or not. The default weight is 0 in original repo.